rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+                                     wildcard: Boolean): NonDetFreeMap[Unit] =
+    (tlist, plist) match {
+      // Handle the remainder.
+      case (rem, Nil) =>
+        varLevel match {
+          case None =>
+            // If there is a wildcard, we succeed.
+            // We also succed if there isn't but the remainder is empty.
+            if (wildcard || rem == Nil)
+              StateT.pure(Unit)
+            else
+              // This should be prevented by the length checks.
+              StateT.liftF(Stream.Empty)
+          // If there's a capture variable, we prefer to add things to that rather than throw them
+          // away.
+          case Some(level) => {
+            // This function is essentially an early terminating left fold.
+            @tailrec
+            def foldRemainder(remainder: List[T], p: Par): NonDetFreeMap[Par] =
+              remainder match {
+                case Nil => StateT.pure(p)
+                case item :: rem =>
+                  if (item.locallyFree.isEmpty)
+                    foldRemainder(rem, merger(p, item))
+                  else if (wildcard)
+                    foldRemainder(rem, p)
+                  else
+                    StateT.liftF(Stream.Empty)
+              }
+            for {
+              p          <- StateT.inspect[Stream, FreeMap, Par]((m: FreeMap) => m.getOrElse(level, Par()))
+              collectPar <- foldRemainder(rem, p)
+              _          <- StateT.modify[Stream, FreeMap]((m: FreeMap) => m + (level -> collectPar))
+            } yield Unit
+          }
+        }
+      // Try to find a match for a single pattern.
+      case (targets, pattern :: prem) => {
+        if (pattern.freeCount == 0) {
+          possiblyRemove(pattern, targets) match {
+            case None           => StateT.liftF(Stream.Empty)
+            case Some(filtered) => listMatch(filtered, prem, matcher, merger, varLevel, wildcard)
+          }
+        } else {
+          for {
+            trem        <- listMatchItem(targets, pattern, matcher)
+            forcedYield <- listMatch(trem, prem, matcher, merger, varLevel, wildcard)
+          } yield forcedYield
+        }
+      }
+    }
+
+  def listMatchItem[T <: HasLocallyFree](
+      tlist: List[T],
+      pattern: T,
+      matcher: (T, T) => OptionalFreeMap[Unit]): NonDetFreeMap[List[T]] =
+    for {
+      triple               <- StateT.liftF(singleOut(tlist).toStream)
+      (head, target, tail) = triple
+      forcedYield <- StateT[Stream, FreeMap, List[T]]((s: FreeMap) => {
+                      matcher(target, pattern).run(s) match {
+                        case None =>
+                          Stream.Empty
+                        case Some((state, _)) =>
+                          Stream((state, head.reverse ++ tail))
+                      }
+                    })
+    } yield forcedYield
+
+  def spatialMatch(target: Par, pattern: Par): OptionalFreeMap[Unit] =
+    if (pattern.freeCount == 0) {
```

Comment:
You can use `===` as `Eq` instance

Reply:
I don't understand this comment. Care to clarify?

Reply:
Sure. `"1" == 1` will compile and always return `false` , `===` is operator from `Eq` typeclass thus `"1" === =` will not compile.

In general I prefer using `Eq` as it give me more type safety.